### PR TITLE
categories and subcategories exercise

### PR DESCRIPTION
**Pull request by agarwalrashika00**
**Category and Subcategory implementation**

  There are many categories, and each category can have many sub_categories.
  Each product belongs to some category(can either belongs to category or sub_category)
  Create db schema and corresponding migrations to implement this in depot app.

  **Validations**
    Category’s name should be mandatory
    All root category names should be unique
    For each category, name of its sub_categories should be unique
    Uniqueness validations should only run if name is present
sub category cannot have any child category. means it is only one level of nesting.

  Create associations in the Category model for the following
    getting all the sub_categories of the category
    getting all the products which belong to category
    getting all the products which belong to subcategories of a given category
each category should have a count column which should have count of total products under it. So parent category should have the count of its products + products belongs to its sub-category.

  **Ensure following things**

    - Should not be able to destroy category if any products are associated with it
    - Should not be able to destroy category if any products are associated with its sub_categories.
    - Destroying a category should destroy all its sub_categories if no products associated     with either of them
Please note that exception should not be raised when calling category.destroy, it should return false in all test cases in which category could not be destroyed. 
On some page display all categories with its sub categories in minimal queries

### CODE REVIEW

File: db/migrate/20230411110435_create_categories.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateCategories < ActiveRecord::Migration[7.0]
+  def change
+    create_table :categories do |t|
+      t.string :name
+      t.bigint :super_category_id
```
Comment (Ankita9818) [inline db/migrate/20230411110435_create_categories.rb:5]: lets try out with references

Comment (agarwalrashika00) [reply db/migrate/20230411110435_create_categories.rb:5]: t.references :super_category, foreign_key: { to_table: :category }

File: app/models/category.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  has_many :sub_categories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+
+  validates :name, presence: true
+  validates :name, uniqueness: { scope: :super_category }, allow_blank: true
+  validate :no_child_of_sub_categories, unless: -> {super_category_id.nil?}
+
+  private
+
+  def no_child_of_sub_categories
+    unless Category.find(super_category_id).super_category_id.nil?
```
Comment (Ankita9818) [inline app/models/category.rb]: lets handle the case if grand child association exists as well

File: app/models/category.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class Category < ApplicationRecord
+  has_many :sub_categories, class_name: "Category", foreign_key: "super_category_id", dependent: :destroy
+  belongs_to :super_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_categories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+
+  validates :name, presence: true
+  validates :name, uniqueness: { scope: :super_category }, allow_blank: true
+  validate :no_child_of_sub_categories, unless: -> {super_category_id.nil?}
```
Comment (Ankita9818) [inline app/models/category.rb]: refactor unless block

File: app/models/product.rb [hunk 4]
```diff
@@ -29,6 +30,10 @@ class Product < ApplicationRecord
 
   scope :enabled, -> { where(enabled: true) }
 
+  after_save :set_count_on_save
```
Comment (Ankita9818) [inline app/models/product.rb:34]: lets discuss

File: app/models/category.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+class Category < ApplicationRecord
+  has_many :sub_categories, class_name: "Category", foreign_key: "super_category_id", dependent: :destroy
+  belongs_to :super_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_categories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+
+  validates :name, presence: true
+  validates :name, uniqueness: { scope: :super_category }, allow_blank: true
```
Comment (agarwalrashika00) [inline app/models/category.rb]: scope: :super_category_id vs scope: :super_category

File: app/models/category.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  has_many :sub_categories, class_name: "Category", foreign_key: "super_category_id", dependent: :destroy
+  belongs_to :super_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_categories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+
+  validates :name, presence: true
+  validates :name, uniqueness: { scope: :super_category_id }, allow_blank: true
+  validate :no_child_of_sub_categories
+
+  after_update_commit :set_products_count, if: :super_category_changed?
```
Comment (agarwalrashika00) [inline app/models/category.rb:11]: super_category_id_changed

File: app/models/product.rb [hunk 7]
```diff
@@ -3,22 +3,24 @@ class Product < ApplicationRecord
...
+  validates :permalink, uniqueness: true, allow_blank: true, format: {
     with: PERMALINK_REGEXP
   }
   validates :description, word_count: { in: 5..10 }, allow_nil: true
   validates_length_of :essay1,
     minimum: 5,
     maximum: 10,
-    message: 'not in range'
+    message: 'not in range',
+    if: -> { essay.present? }
```
Comment (Ankita9818) [inline app/models/product.rb:23]: essay?

**Review by Ankita9818 — APPROVED** [review]

**Pull request merged**